pkg/compactor/sharded_block_populator.go

```diff
@@ -0,0 +1,217 @@
+package compactor
+
+import (
+	"context"
+	"io"
+	"maps"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"golang.org/x/exp/slices"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+)
+
+type ShardedBlockPopulator struct {
+	partitionCount int
+	partitionID    int
+	logger         log.Logger
+}
+
+// PopulateBlock fills the index and chunk writers with new data gathered as the union
+// of the provided blocks. It returns meta information for the new block.
+// It expects sorted blocks input by mint.
+// The main logic is copied from tsdb.DefaultPopulateBlockFunc
+func (c ShardedBlockPopulator) PopulateBlock(ctx context.Context, metrics *tsdb.CompactorMetrics, _ log.Logger, chunkPool chunkenc.Pool, mergeFunc storage.VerticalChunkSeriesMergeFunc, blocks []tsdb.BlockReader, meta *tsdb.BlockMeta, indexw tsdb.IndexWriter, chunkw tsdb.ChunkWriter, postingsFunc tsdb.IndexReaderPostingsFunc) (err error) {
+	if len(blocks) == 0 {
+		return errors.New("cannot populate block from no readers")
+	}
+
+	var (
+		sets        []storage.ChunkSeriesSet
+		setsMtx     sync.Mutex
+		symbols     map[string]struct{}
+		closers     []io.Closer
+		overlapping bool
+	)
+	symbols = make(map[string]struct{})
+	defer func() {
+		errs := tsdb_errors.NewMulti(err)
+		if cerr := tsdb_errors.CloseAll(closers); cerr != nil {
+			errs.Add(errors.Wrap(cerr, "close"))
+		}
+		err = errs.Err()
+		metrics.PopulatingBlocks.Set(0)
+	}()
+	metrics.PopulatingBlocks.Set(1)
+
+	globalMaxt := blocks[0].Meta().MaxTime
+	g, _ := errgroup.WithContext(ctx)
+	g.SetLimit(8)
+	for i, b := range blocks {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		if !overlapping {
+			if i > 0 && b.Meta().MinTime < globalMaxt {
+				metrics.OverlappingBlocks.Inc()
+				overlapping = true
+				level.Info(c.logger).Log("msg", "Found overlapping blocks during compaction", "ulid", meta.ULID)
+			}
+			if b.Meta().MaxTime > globalMaxt {
+				globalMaxt = b.Meta().MaxTime
+			}
```

Comment:
Although it is part of the original tsdb implementation and we also pass tsdb metrics in the function. I feel it is kind of unecessary to check block overlapping for partitioning compactor. Especially the info log above seems always there

Reply:
Make sense. I will remove that part.